File: src/chapters/linear-systems/content/code/gauss-complete.m
function x = gauss_complete_pivot(A, b)
% GAUSS_COMPLETE_PIVOT  Gaussian elimination with complete (row+col) pivoting.
    b = b(:); n = numel(b); col = 1:n;
    for k = 1:n
        sub = abs(A(k:n, k:n));
        [~, idx] = max(sub(:));
        [i, j] = ind2sub(size(sub), idx); i = i + k - 1; j = j + k - 1;
        A([k i], :) = A([i k], :); b([k i]) = b([i k]);
        A(:, [k j]) = A(:, [j k]); col([k j]) = col([j k]);
        for r = k+1:n
            f = A(r,k) / A(k,k);
            A(r, k:n) = A(r, k:n) - f * A(k, k:n);
            b(r) = b(r) - f * b(k);
        end
    end
    y = zeros(n, 1);
    for i = n:-1:1
        y(i) = (b(i) - A(i, i+1:n) * y(i+1:n)) / A(i, i);
    end
    x = zeros(n, 1); x(col) = y;                          % undo column swaps
end

% --- Demo ---
A = [2 1 -1; -3 -1 2; -2 1 2]; b = [8; -11; -3];
disp(gauss_complete_pivot(A, b)');

File: src/chapters/linear-systems/content/code/gauss-complete.r
gauss_complete_pivot <- function(A, b) {
  # Gaussian elimination with complete (row+column) pivoting.
  # Returns x (column swaps are undone before returning).
  A <- matrix(as.numeric(A), nrow = nrow(A))
  b <- as.numeric(b)
  n <- length(b)
  col <- 1:n                                 # tracks column permutation
  for (k in 1:n) {
    sub <- abs(A[k:n, k:n, drop = FALSE])
    idx <- which.max(sub)                    # column-major linear index
    i <- ((idx - 1) %% nrow(sub)) + k
    j <- ((idx - 1) %/% nrow(sub)) + k
    A[c(k, i), ] <- A[c(i, k), ]; b[c(k, i)] <- b[c(i, k)]
    A[, c(k, j)] <- A[, c(j, k)]; col[c(k, j)] <- col[c(j, k)]
    if (k < n) {
      for (r in (k + 1):n) {
        f <- A[r, k] / A[k, k]
        A[r, k:n] <- A[r, k:n] - f * A[k, k:n]
        b[r] <- b[r] - f * b[k]
      }
    }
  }
  y <- numeric(n)
  for (i in n:1) {
    s <- b[i]
    if (i < n) s <- s - sum(A[i, (i + 1):n] * y[(i + 1):n])
    y[i] <- s / A[i, i]
  }
  x <- numeric(n)
  x[col] <- y                                # undo column permutation
  x
}

A <- matrix(c( 2,  1, -1,
              -3, -1,  2,
              -2,  1,  2), nrow = 3, byrow = TRUE)
b <- c(8, -11, -3)
cat("x =", gauss_complete_pivot(A, b), "\n")
